knimin/lib/data_access.py

```diff
@@ -3894,6 +3894,95 @@ def delete_shotgun_index_aliquot(self, shotgun_index_aliquot_id):
+            A list of str of string pairs (i5, i7) depending if the indexing
+            technology is dual barcoded or not
+
+        Notes
+        -----
+            This generates a list of i7 barcodes or the combination of
+            i5 and i7 barcodes. The combinations are more complex as we need
+            to make sure that we have actual rotation of those combinations.
+            The general formula we will be using for index selection is:
+            index_num = sample_number % len-availavle-idx + (
+                sample_number / len-availavle-idx) % len-availavle-idx
+        """
+        with TRN:
+            sql = """SELECT shotgun_index_tech_id, name, dual_index,
+                        last_index_idx
+                     FROM pm.shotgun_index_tech WHERE name = %s"""
+            TRN.add(sql, [idx_tech])
+            idx_tech_vals = TRN.execute_fetchindex()
+
+            if not idx_tech_vals:
+                raise ValueError('Not a valid shotgun index '
+                                 'technology: %s' % idx_tech)
+            else:
+                # just taking the first element and converting it to dict
+                idx_tech_vals = dict([v for v in idx_tech_vals][0])
+
+            # selecting single indexing
+            sql = """SELECT shotgun_index_id
+                     FROM pm.shotgun_index
+                     WHERE shotgun_index_tech_id = %s AND index_type = 'i7'
+                     ORDER BY shotgun_index_id"""
+            TRN.add(sql, [idx_tech_vals['shotgun_index_tech_id']])
+            indices_i7 = [r[0] for r in TRN.execute_fetchindex()]
+            if not indices_i7:
+                raise ValueError("%s doesn't have any i7 values" % idx_tech)
+            len_indices = len(indices_i7)
+
+            if idx_tech_vals['dual_index']:
+                # selecting the dual section
+                sql = """SELECT shotgun_index_id
+                         FROM pm.shotgun_index
+                         WHERE shotgun_index_tech_id = %s AND index_type = 'i5'
+                         ORDER BY shotgun_index_id"""
+                TRN.add(sql, [idx_tech_vals['shotgun_index_tech_id']])
+                indices_i5 = [r[0] for r in TRN.execute_fetchindex()]
+                if not indices_i5:
+                    raise ValueError("%s doesn't have any i5 "
+                                     "values" % idx_tech)
+
+                len_indices = min([len_indices, len(indices_i5)])
+
+            idx_start = idx_tech_vals['last_index_idx']
+            indices_generated = []
+            for i in range(1, num_samples + 1):
+                sample_number = idx_start + i
+                index_num = (sample_number % len_indices +
+                             (sample_number / len_indices) % len_indices)
+
+                if idx_tech_vals['dual_index']:
+                    indices_generated.append((indices_i5[index_num],
+                                              indices_i7[index_num]))
```

Comment:
I did some more thinking about this, and I have a different suggested solution:

```python
def pick_barcodes(i5, i7, num_samples, start=0):
    samples = {}
    len_i5 = len(i5)
    len_i7 = len(i7)
    for i in range(start,start+num_samples):
        sample_i5 = i % len_i5
        sample_i7 = int(((i % len_i5) + floor(float(i) / float(len_i5))) % len_i7)
        samples[i] = (i5[sample_i5], i7[sample_i7])
    
    return(samples)
```

I think that this gives the correct logic despite different numbers of i5 and i7 barcodes:

```python
In [9]: nextera_i5 = ['D501','D502','D503','D504','D505','D506','D507','D508']

In [10]: nextera_i7 = ['D701','D702','D703','D704','D705','D706','D707','D708','D709','D710','D711','D712']

In [11]: combos = pick_barcodes(nextera_i5, nextera_i7, 96)

In [12]: len(set(combos.values()))
Out[12]: 96

In [13]: combos = pick_barcodes(nextera_i5, nextera_i7, 97)

In [14]: len(set(combos.values()))
Out[14]: 96
```

This scales to larger numbers of barcodes:
```python
In [15]: iTru_i5 = range(48)

In [16]: iTru_i7 = range(48)

In [17]: combos = pick_barcodes(iTru_i5, iTru_i7, 97)

In [18]: len(set(combos.values()))
Out[18]: 97

In [19]: combos = pick_barcodes(iTru_i5, iTru_i7, 48*48)

In [20]: len(set(combos.values()))
Out[20]: 2304

In [21]: combos = pick_barcodes(iTru_i5, iTru_i7, 48*48 + 1)

In [22]: len(set(combos.values()))
Out[22]: 2304
```

The original modulo math I quoted you is wrong I think:
```python
def orig_strategy(i5, i7, num_samples, start=0):
    len_indices = min([len(i7), len(i5)])
    
    samples = {}
    indices_generated = []
    for i in range(start,start+num_samples):
        index_num = (i % len_indices +
                    (i / len_indices) % len_indices)    
        samples[i] = (i5[index_num], i7[index_num])
    
    return(samples)
```

It reuses barcode combos too quickly:
```python
In [24]: combos = orig_strategy(nextera_i5, nextera_i7, 8)

In [25]: len(set(combos.values()))
Out[25]: 8

In [26]: combos = orig_strategy(nextera_i5, nextera_i7, 12)

In [27]: len(set(combos.values()))
Out[27]: 8
```

and it breaks:
```python
In [28]: combos = orig_strategy(nextera_i5, nextera_i7, 16)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-28-8429db63caea> in <module>()
----> 1 combos = orig_strategy(nextera_i5, nextera_i7, 16)

<ipython-input-23-a7f3be7d424f> in orig_strategy(i5, i7, num_samples, start)
      7         index_num = (i % len_indices +
      8                     (i / len_indices) % len_indices)    
----> 9         samples[i] = (i5[index_num], i7[index_num])
     10 
     11     return(samples)

IndexError: list index out of range
```